libs/events/events_test.go

```diff
@@ -196,11 +179,27 @@ func TestAddAndRemoveListenerConcurrency(t *testing.T) {
 	}
 	defer evsw.Stop()
 
-	done1 := make(chan uint64)
-	done2 := make(chan uint64)
+	done1 := make(chan struct{})
+	done2 := make(chan struct{})
+
+	go func() {
+		for i := 0; i < roundCount; i++ {
+			evsw.RemoveListener("listener")
+		}
+		done1 <- struct{}{}
+	}()
 
-	go inputRemoveListener(t, evsw, done1, roundCount)
-	go inputAddListenerForEvent(t, evsw, done2, roundCount, &stopInputEvent)
+	go func() {
+		for i := 0; i < roundCount; i++ {
+			index := i //it necessary for closure
+			evsw.AddListenerForEvent("listener", fmt.Sprintf("event%d", index),
+				func(data EventData) {
+					t.Errorf("should not run callback for %d.\n", index)
+					stopInputEvent = true
+				})
+		}
+		done2 <- struct{}{}
```

Comment:
`close(done2)`

Reply:
here chan can recycled automatically by the garbage collector， but I'll close it explicit